File: src/getsomepuzzle/engine/constraints/parity.py
import random

from ..utils import to_rows, to_columns
from .base import CellCentricConstraint
from ..constants import CONSTRAST, EMPTY


class ParityConstraint(CellCentricConstraint):
    slug = "PA"

    def __repr__(self):
        indices, side = self.parameters["indices"], self.parameters["side"]
        idx = indices[0]
        return (
            f"Cell {idx + 1} should have the same number "
            f"of odd and even numbers on its {side} "
            f"side{'s' if side in ('horizontal', 'vertical') else ''}"
        )

    def conflicts(self, other):
        if not isinstance(other, ParityConstraint):
            return False

        # Parity only have one index
        if self.parameters["indices"][0] != other.parameters["indices"][0]:
            return False

        conflicting = {
            "left": ["horizontal", "right"],
            "right": ["horizontal", "left"],
            "top": ["vertical", "bottom"],
            "bottom": ["vertical", "top"],
            "horizontal": ["right", "left"],
            "vertical": ["bottom", "top"],
        }
        return other.parameters["side"] in conflicting[self.parameters["side"]]

    def get_cell_text(self):
        # ⬅ ⮕ ⬆ ⬇ ⬌ ⬍  ⬉ ⬈ ⬊ ⬋
        parity_icons = {
            "left": "⬅",
            "right": "⮕",
            "horizontal": "⬌",
            "top": "⬆",
            "bottom": "⬇",
            "vertical": "⬍",
        }
        return parity_icons[self.parameters["side"]]

    def _check(self, puzzle, debug=False):
        # The given cell has the same number of odd and even numbers on
        # one of its side (or both, vertically or horizontally),
        # the given cell and side are defined in the parameters
        indices, side = self.parameters["indices"], self.parameters["side"]
        idx = indices[0]
        w, h = puzzle.width, puzzle.height
        # Find the right row for idx
        ridx = idx // w
        cidx = idx % w
        rows = to_rows(puzzle.state, w, h)
        row = rows[ridx]
        columns = to_columns(puzzle.state, w)
        column = columns[cidx]
        row_values_and_indices = [(idx, c.value) for idx, c in enumerate(row)]
        col_values_and_indices = [(idx, c.value) for idx, c in enumerate(column)]
        sides = []
        if side in ("left", "horizontal"):
            sides.append([v for (i, v) in row_values_and_indices if i < cidx])
        if side in ("right", "horizontal"):
            sides.append([v for (i, v) in row_values_and_indices if i > cidx])
        if side in ("top", "vertical"):
            sides.append([v for (i, v) in col_values_and_indices if i < ridx])
        if side in ("bottom", "vertical"):
            sides.append([v for (i, v) in col_values_and_indices if i > ridx])
        for side in sides:
            if any(v == EMPTY for v in side):
                continue
            even = len([v for v in side if v % 2 == 0])
            odd = len([v for v in side if v % 2 != 0])
            if even != odd:
                return False
        return True

    def apply(self, puzzle):
        changed = False
        indices, side = self.parameters["indices"], self.parameters["side"]
        idx = indices[0]
        w, h = puzzle.width, puzzle.height
        ridx = idx // w
        cidx = idx % w
        if side in ("left", "right", "horizontal"):
            rows = to_rows(puzzle.state, w, h)
            row = rows[ridx]
            values_and_indices = [(idx, ridx, c.value) for idx, c in enumerate(row)]
        else:
            columns = to_columns(puzzle.state, w)
            column = columns[cidx]
            values_and_indices = [(cidx, idx, c.value) for idx, c in enumerate(column)]

        sides = []
        if side in ("left", "horizontal"):
            sides.append([(c, r, v) for (c, r, v) in values_and_indices if c < cidx])
        if side in ("right", "horizontal"):
            sides.append([(c, r, v) for (c, r, v) in values_and_indices if c > cidx])
        if side in ("top", "vertical"):
            sides.append([(c, r, v) for (c, r, v) in values_and_indices if r < ridx])
        if side in ("bottom", "vertical"):
            sides.append([(c, r, v) for (c, r, v) in values_and_indices if r > ridx])
        for side in sides:
            empty_cells = [v == EMPTY for (c, r, v) in side]
            if not any(empty_cells):
                continue
            if all(empty_cells):
                continue
            even = len([v for (c, r, v) in side if v != EMPTY and v % 2 == 0])
            odd = len([v for (c, r, v) in side if v!= EMPTY and v % 2 != 0])
            empty = empty_cells.count(True)
            size_per_parity = int(len(side) / 2)
            value = None
            if size_per_parity - even == empty:
                value = [v for v in puzzle.domain if v % 2 == 0][0]
            elif size_per_parity - odd == empty:
                value = [v for v in puzzle.domain if v % 2 != 0][0]
            if value is not None:
                for (c, r, v) in side:
                    if v != EMPTY:
                        continue
                    idx = r * w + c
                    changed |= puzzle.state[idx].set_value(value)
        return changed

    @staticmethod
    def generate_random_parameters(puzzle):
        # { "indices" : [4], "side" : "left" or "horizontal" or "right" }
        width = puzzle.width
        height = puzzle.height
        choices = ["left", "right", "top", "bottom"]
        if width % 2 != 0:
            choices.append("horizontal")
        if height % 2 != 0:
            choices.append("vertical")
        side = random.choice(choices)
        if side in ("left", "right", "horizontal"):
            possible_indices = [i for i in range(width)]
            if side in ("left", "horizontal"):
                possible_indices = [i for i in possible_indices if i >= 2 and i % 2 == 0]
            if side in ("right", "horizontal"):
                possible_indices = [i for i in possible_indices if i <= width - 2 and (width - 1 - i) % 2 == 0]
            if not possible_indices:
                raise ValueError("Cannot generate parity constraint")
            col = random.choice(possible_indices)
            row = random.randint(0, height - 1)
            return {"indices": [col + row * puzzle.width], "side": side}

        else:
            possible_indices = [i for i in range(height)]
            if side in ("top", "vertical"):
                possible_indices = [i for i in possible_indices if i >= 2 and i % 2 == 0]
            if side in ("bottom", "vertical"):
                possible_indices = [i for i in possible_indices if i <= height - 2 and (height - 1 - i) % 2 == 0]
            if not possible_indices:
                raise ValueError("Cannot generate parity constraint")
            col = random.randint(0, puzzle.width - 1)
            row = random.choice(possible_indices)
            return {"indices": [col + row * puzzle.width], "side": side}

    @staticmethod
    def generate_all_parameters(puzzle):
        w = puzzle.width
        h = puzzle.height
        for idx in range(len(puzzle.state)):
            ridx = idx // w
            cidx = idx % w
            left_size = cidx
            right_size = w - 1 - cidx
            top_size = ridx
            bottom_size = h - 1 - ridx
            if left_size % 2 == 0 and left_size > 0:
                yield {"indices": [idx], "side": "left"}
            if right_size % 2 == 0 and right_size > 0:
                yield {"indices": [idx], "side": "right"}
            if left_size % 2 == 0 and right_size % 2 == 0 and right_size > 0 and left_size > 0:
                yield {"indices": [idx], "side": "horizontal"}
            if top_size % 2 == 0 and top_size > 0:
                yield {"indices": [idx], "side": "top"}
            if bottom_size % 2 == 0 and bottom_size > 0:
                yield {"indices": [idx], "side": "bottom"}
            if top_size % 2 == 0 and bottom_size % 2 == 0 and bottom_size > 0 and top_size > 0:
                yield {"indices": [idx], "side": "vertical"}


    @staticmethod
    def maximum_presence(puzzle):
        return int((puzzle.width * puzzle.height) / 5)

    def line_export(self):
        indices, side = self.parameters["indices"], self.parameters["side"]
        idx = indices[0]
        return f"{self.slug}:{idx}.{side}"

    @staticmethod
    def line_import(line):
        idx, side = line.split(".")
        return {"indices": [int(idx)], "side": side}

    def signature(self):
        bilateral = "bi" if self.parameters["side"] in ("horizontal", "vertical") else "mo"
        return f"{self.slug}:{bilateral}"

File: lib/getsomepuzzle/constraints/parity.dart
import 'package:getsomepuzzle/getsomepuzzle/model/cell.dart';
import 'package:getsomepuzzle/getsomepuzzle/constraints/constraint.dart';
import 'package:getsomepuzzle/getsomepuzzle/model/puzzle.dart';
import 'package:getsomepuzzle/getsomepuzzle/utils/rotation.dart';

/// `PA` — **balanced colour partition** of one side of an anchor cell.
///
/// Despite the legacy "parity" name and slug, the constraint requires every
/// colour of the puzzle's domain to appear the *same* number of times on the
/// targeted side: `targetCount = side.length / domain.length` cells per colour
/// (see [verify] / [apply]). The name is historical — on the original
/// 2-colour domain "balanced" collapses to "as many black as white", i.e. an
/// even side split in half, which reads as parity. On a 3-colour domain it
/// generalises to "as many black as white as purple", which is no longer a
/// parity statement; the slug is kept for backward compatibility with stored
/// puzzle lines. See `docs/dev/third_color.md` § "PA semantics on 3 colours".
class ParityConstraint extends CellsCentricConstraint {
  @override
  String get slug => 'PA';

  // Colour-agnostic: balances colours per side, references none specifically.
  @override
  Set<CellValue> get referencedColors => const {};

  String side = "";

  ParityConstraint(String strParams) {
    indices.add(int.parse(strParams.split(".")[0]));
    side = strParams.split(".")[1];
  }

  @override
  String toString() {
    if (side == "left") return "⬅";
    if (side == "right") return "⮕";
    if (side == "horizontal") return "⬌";
    if (side == "vertical") return "⬍";
    if (side == "top") return "⬆";
    if (side == "bottom") return "⬇";
    return "";
  }

  @override
  String serialize() => 'PA:${indices.first}.$side';

  /// Returns the merged side when [a] and [b] lie on the same axis
  /// (vertical = top/bottom, horizontal = left/right), null otherwise.
  /// Each side of a `vertical`/`horizontal` constraint must be balanced
  /// independently, so `vertical` subsumes `top` and `bottom` (and the
  /// pair `top` + `bottom` is exactly `vertical`) — same for the
  /// horizontal axis.
  static String? mergeSides(String a, String b) {
    if (a == b) return a;
    const vertical = {'top', 'bottom', 'vertical'};
    const horizontal = {'left', 'right', 'horizontal'};
    if (vertical.contains(a) && vertical.contains(b)) return 'vertical';
    if (horizontal.contains(a) && horizontal.contains(b)) return 'horizontal';
    return null;
  }

  /// Rotation-90°-CW mapping for the `side` parameter. Cells originally to
  /// the left of the anchor end up above it in the rotated grid, etc.
  static const Map<String, String> _rotatedSide = {
    'left': 'top',
    'right': 'bottom',
    'top': 'right',
    'bottom': 'left',
    'horizontal': 'vertical',
    'vertical': 'horizontal',
  };

  @override
  Constraint rotated(int origWidth, int origHeight) {
    final newIdx = rotateIdx90CW(indices.first, origWidth, origHeight);
    final newSide = _rotatedSide[side] ?? side;
    return ParityConstraint('$newIdx.$newSide');
  }

  @override
  String toHuman(Puzzle puzzle) {
    final idx = indices.first + 1;
    return "$idx = ${toString()}";
  }

  /// Generate all valid parity constraint parameters for a given grid size.
  ///
  /// A side is only emitted when its length is divisible by `domain.length`,
  /// so the per-colour target `side.length / domain.length` is an integer.
  /// On a 2-colour domain this is the classic even-side test (`size % 2`); on
  /// 3 colours it becomes `size % 3`, etc.
  static List<String> generateAllParameters(
    int width,
    int height,
    List<CellValue> domain,
    Set<int>? excludedIndices,
  ) {
    final List<String> result = [];
    for (int idx = 0; idx < width * height; idx++) {
      final ridx = idx ~/ width;
      final cidx = idx % width;
      final leftSize = cidx;
      final rightSize = width - 1 - cidx;
      final topSize = ridx;
      final bottomSize = height - 1 - ridx;
      if (leftSize % domain.length == 0 && leftSize > 0) {
        result.add('$idx.left');
      }
      if (rightSize % domain.length == 0 && rightSize > 0) {
        result.add('$idx.right');
      }
      if (leftSize % domain.length == 0 &&
          rightSize % domain.length == 0 &&
          rightSize > 0 &&
          leftSize > 0) {
        result.add('$idx.horizontal');
      }
      if (topSize % domain.length == 0 && topSize > 0) {
        result.add('$idx.top');
      }
      if (bottomSize % domain.length == 0 && bottomSize > 0) {
        result.add('$idx.bottom');
      }
      if (topSize % domain.length == 0 &&
          bottomSize % domain.length == 0 &&
          bottomSize > 0 &&
          topSize > 0) {
        result.add('$idx.vertical');
      }
    }
    return result;
  }

  List<List<CellValue>> _getSideValues(Puzzle puzzle) {
    return _getSideCells(
      puzzle,
    ).map((side) => side.map((e) => e.$2.value).toList()).toList();
  }

  List<List<(int, Cell)>> _getSideCells(Puzzle puzzle) {
    final w = puzzle.width;
    final idx = indices[0];
    final ridx = idx ~/ w;
    final cidx = idx % w;
    final rows = puzzle.getRows();
    final row = rows[ridx];
    final columns = puzzle.getColumns();
    final column = columns[cidx];
    final rowValuesAndCells = row.indexed;
    final colValuesAndCells = column.indexed;
    final List<List<(int, Cell)>> sides = [];
    if (side == "left" || side == "horizontal") {
      sides.add(rowValuesAndCells.where((e) => e.$1 < cidx).toList());
    }
    if (side == "right" || side == "horizontal") {
      sides.add(rowValuesAndCells.where((e) => e.$1 > cidx).toList());
    }
    if (side == "top" || side == "vertical") {
      sides.add(colValuesAndCells.where((e) => e.$1 < ridx).toList());
    }
    if (side == "bottom" || side == "vertical") {
      sides.add(colValuesAndCells.where((e) => e.$1 > ridx).toList());
    }
    return sides;
  }

  @override
  bool verify(Puzzle puzzle) {
    // Use cells (not just values) so reachability can account for free
    // cells that have pruned a colour from their options.
    for (var side in _getSideCells(puzzle)) {
      final int targetCount = side.length ~/ puzzle.domain.length;
      final bool hasFree = side.any((e) => e.$2.value == CellValue.free);
      for (var color in puzzle.domain) {
        final int have = side.where((e) => e.$2.value == color).length;
        // Too many of one color already → target is
        // unreachable from this state
        if (have > targetCount) return false;
        // For a fully-filled side the counts must match exactly
        if (!hasFree && have != targetCount) return false;
        // Reachability: even using every free cell that still allows this
        // colour, the side can't reach its per-colour target → unreachable.
        final int reachableFree = side
            .where(
              (e) =>
                  e.$2.value == CellValue.free && e.$2.options.contains(color),
            )
            .length;
        if (have + reachableFree < targetCount) return false;
      }
    }
    return true;
  }

  @override
  Move? apply(Puzzle puzzle) {
    final sides = _getSideCells(puzzle);
    // Weight by the largest side covered: with 2 cells per side, the second
    // cell is read directly off the first; 4 cells require parity counting;
    // 6+ cells need real bookkeeping. For 2-side variants (horizontal /
    // vertical) we take the max — the player must still scan the long side.
    final int maxSide = sides
        .map((s) => s.length)
        .reduce((a, b) => a > b ? a : b);
    final int weight = maxSide <= puzzle.domain.length
        ? 0
        : (maxSide <= (puzzle.domain.length * 2) ? 1 : 2);
    for (var side in sides) {
      final int targetCount = side.length ~/ puzzle.domain.length;
      final freeCells = side.where(
        (element) => element.$2.value == CellValue.free,
      );
      final Map<CellValue, int> perColor = {};
      for (var color in puzzle.domain) {
        perColor[color] = side
            .where((v) => v.$2.value != CellValue.free && v.$2.value == color)
            .length;
        // If we're already above the target, it's an "isImpossible" move
        if (perColor[color]! > targetCount) {
          return Impossible(this);
        }
        // If this color has reached its target, all the free cells that still
        // have that option must remove it. Removing the last remaining option
        // has the side effect of applying the only remaining color.
        if (perColor[color]! == targetCount) {
          for (var freeCell in freeCells) {
            if (freeCell.$2.options.contains(color)) {
              return RemoveOption(
                freeCell.$2.idx,
                color,
                this,
                complexity: weight,
              );
            }
          }
        }
      }
    }
    return null;
  }

  @override
  bool isCompleteFor(Puzzle puzzle) {
    if (!verify(puzzle)) return false;
    final sides = _getSideValues(puzzle);
    for (var side in sides) {
      if (side.contains(CellValue.free)) return false;
    }
    return true;
  }
}
